### ISSUE TITLE
scanpy.pl.spatial does not work properly in plt.subplots

### ISSUE THREAD

**Issue by linanzhang** [scverse/scanpy#1158]
I have spatial annotations for multiple datasets and want to put them in the same plt.subplots framework. Only the last one gets the markers.

```python
fig, ax = plt.subplots(1,3, figsize=(20,6))
sc.pl.spatial(adata, img_key="hires", color="clusters", size=1.5, ax=ax[0], show=False)
sc.pl.spatial(bdata, img_key="hires", color="clusters", size=1.5, ax=ax[1], show=False)
sc.pl.spatial(cdata, img_key="hires", color="clusters", size=1.5, ax=ax[2], show=False)
plt.tight_layout(pad=3.0)
plt.show()
```

![download](https://user-images.githubusercontent.com/37935731/78901496-64765f00-7a46-11ea-83ba-56dc16ff5e14.png)

**Comment by vitkl** [comment]
Same problem with subplots here but I use the same adata object, I plot different columns with `color` and I need to use a different `size` for each column. I get empty subplot panels but one and the rest of plots are shown and individual plots below the panels.

**Comment by giovp** [comment]
Sorry for late reply, I think this was fixed in #1138. Could you update your scanpy and try again?
For me it seems to work
```python
fig, ax = plt.subplots(1,3, figsize=(20,6))
sc.pl.spatial(adata, img_key="hires", color="array_row", size=1.5, ax=ax[0], show=False)
sc.pl.spatial(bdata, img_key="hires", color="array_row", size=1.5, ax=ax[1], show=False)
sc.pl.spatial(cdata, img_key="hires", color="array_row", size=1.5, ax=ax[2], show=False)
plt.tight_layout(pad=3.0)
plt.show()
```
![image](https://user-images.githubusercontent.com/25887487/79438766-41165b80-7fd4-11ea-8ed7-f297b22da7c0.png)

**Comment by linanzhang** [comment]
I updated scanpy:

```
pip install git+https://github.com/theislab/scanpy.git
pip install spatialde
```


```
Successfully built scanpy
Installing collected packages: scanpy
  Attempting uninstall: scanpy
    Found existing installation: scanpy 1.4.5.2.dev38+gae88b949
    Uninstalling scanpy-1.4.5.2.dev38+gae88b949:
      Successfully uninstalled scanpy-1.4.5.2.dev38+gae88b949
Successfully installed scanpy-1.4.7.dev47+gf6a49e81
```

Now I'm not able to generate a spatial plot at all.

```
----> 5 sc.pl.spatial(adata, img_key="hires", color="clusters", size=1.5)

~/opt/anaconda3/envs/scanpy/lib/python3.8/site-packages/scanpy/plotting/_tools/scatterplots.py in spatial(adata, img_key, library_id, crop_coord, alpha_img, bw, **kwargs)
    763     """
    764     if library_id is _empty:
--> 765         library_id = next((i for i in adata.uns['spatial'].keys()))
    766     else:
    767         if library_id not in adata.uns['spatial'].keys():

KeyError: 'spatial'

```

**Comment by linanzhang** [comment]
And there are more issues with reinstalling scanpy:
`adata = sc.datasets.visium_sge('V1_Breast_Cancer_Block_A_Section_1')`

```
NameError                                 Traceback (most recent call last)
~/opt/anaconda3/envs/scanpy/lib/python3.7/site-packages/tqdm/notebook.py in status_printer(_, total, desc, ncols)
     97             else:  # No total? Show info style bar with no progress tqdm status
---> 98                 pbar = IProgress(min=0, max=1)
     99                 pbar.value = 1

NameError: name 'IProgress' is not defined

During handling of the above exception, another exception occurred:

ImportError                               Traceback (most recent call last)
<ipython-input-11-27e5eaf7b8a6> in <module>
----> 1 adata = sc.datasets.visium_sge('V1_Breast_Cancer_Block_A_Section_1')

~/opt/anaconda3/envs/scanpy/lib/python3.7/site-packages/scanpy/datasets/_datasets.py in visium_sge(sample_id)
    361     Annotated data matrix.
    362     """
--> 363     _download_visium_dataset(sample_id)
    364     return read_visium(settings.datasetdir / sample_id)

~/opt/anaconda3/envs/scanpy/lib/python3.7/site-packages/scanpy/datasets/_datasets.py in _download_visium_dataset(sample_id, base_dir)
    320     tar_pth = sample_dir / tar_filename
    321     _utils.check_presence_download(
--> 322         filename=tar_pth, backup_url=url_prefix + tar_filename
    323     )
    324     with tarfile.open(tar_pth) as f:

~/opt/anaconda3/envs/scanpy/lib/python3.7/site-packages/scanpy/_utils.py in check_presence_download(filename, backup_url)
    589         from .readwrite import _download
    590 
--> 591         _download(backup_url, filename)
    592 
    593 

~/opt/anaconda3/envs/scanpy/lib/python3.7/site-packages/scanpy/readwrite.py in _download(url, path)
    895 
    896     path.parent.mkdir(parents=True, exist_ok=True)
--> 897     with tqdm(unit='B', unit_scale=True, miniters=1, desc=path.name) as t:
    898 
    899         def update_to(b=1, bsize=1, tsize=None):

~/opt/anaconda3/envs/scanpy/lib/python3.7/site-packages/tqdm/notebook.py in __init__(self, *args, **kwargs)
    207         total = self.total * unit_scale if self.total else self.total
    208         self.container = self.status_printer(
--> 209             self.fp, total, self.desc, self.ncols)
    210         self.sp = self.display
    211 

~/opt/anaconda3/envs/scanpy/lib/python3.7/site-packages/tqdm/notebook.py in status_printer(_, total, desc, ncols)
    102             # #187 #451 #558
    103             raise ImportError(
--> 104                 "FloatProgress not found. Please update jupyter and ipywidgets."
    105                 " See https://ipywidgets.readthedocs.io/en/stable"
    106                 "/user_install.html")

ImportError: FloatProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
```

All I did was creating a new environment and installing scanpy following [this](https://github.com/theislab/scanpy-tutorials/blob/master/analysis-visualization-spatial.ipynb).

**Comment by linanzhang** [comment]
Okay I figured it out. You changed the key 'X_spatial' to 'coords' in anndata objects in the new version, which is not compatible with my previously saved .h5ad files.

**Comment by giovp** [comment]
Yes exactly, the structure of `adata.uns` has changed so you might have to import that visium dataset again and reprocess it. Sorry for that, should have mentioned it before.

**Comment by vitkl** [comment]
Where can I find the description of how it changed? Does it support
multiple samples now?
Thanks!

**Comment by giovp** [comment]
You can have a look at the new structure in the function definition [here](https://github.com/theislab/scanpy/blob/4156314407c5368fa0b66ac18470d80f3748a71f/scanpy/readwrite.py#L281) . 

Regarding multiple samples support, that is on its way! Will post here as soon as it's ready and tested.

**Comment by giovp** [comment]
@vitkl now multiple samples are supported, see [here](https://scanpy-tutorials.readthedocs.io/en/latest/spatial/integration-scanorama.html) for description on how to use the new concat strategy

**Comment by vitkl** [comment]
Nice! Thanks!

**Comment by xiao-kong-long** [comment]
> Sorry for late reply, I think this was fixed in #1138. Could you update your scanpy and try again? For me it seems to work
> 
> ```python
> fig, ax = plt.subplots(1,3, figsize=(20,6))
> sc.pl.spatial(adata, img_key="hires", color="array_row", size=1.5, ax=ax[0], show=False)
> sc.pl.spatial(bdata, img_key="hires", color="array_row", size=1.5, ax=ax[1], show=False)
> sc.pl.spatial(cdata, img_key="hires", color="array_row", size=1.5, ax=ax[2], show=False)
> plt.tight_layout(pad=3.0)
> plt.show()
> ```
> 
> ![image](https://user-images.githubusercontent.com/25887487/79438766-41165b80-7fd4-11ea-8ed7-f297b22da7c0.png)

Hello, I have a problem, that is why some plots show colorbar but other plots show legend? It seems using same code.

**Comment by xiao-kong-long** [comment]
> > Sorry for late reply, I think this was fixed in #1138. Could you update your scanpy and try again? For me it seems to work
> > ```python
> > fig, ax = plt.subplots(1,3, figsize=(20,6))
> > sc.pl.spatial(adata, img_key="hires", color="array_row", size=1.5, ax=ax[0], show=False)
> > sc.pl.spatial(bdata, img_key="hires", color="array_row", size=1.5, ax=ax[1], show=False)
> > sc.pl.spatial(cdata, img_key="hires", color="array_row", size=1.5, ax=ax[2], show=False)
> > plt.tight_layout(pad=3.0)
> > plt.show()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ![image](https://user-images.githubusercontent.com/25887487/79438766-41165b80-7fd4-11ea-8ed7-f297b22da7c0.png)
> 
> Hello, I have a problem, that is why some plots show colorbar but other plots show legend? It seems using same code.

I've got the reason: if the data type is category, sc.pl.spetial would append lengend, in others conditon it would append colorbar.
